cmd/mysql-state-monitor/main.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	state "github.com/percona/percona-server-mysql-operator/cmd/internal/naming"
+	"github.com/percona/percona-server-mysql-operator/pkg/naming"
+)
+
+func parseDatum(datum string) state.MySQLState {
+	lines := strings.Split(datum, "\n")
+
+	if lines[0] == "READY=1" {
+		return state.MySQLReady
+	}
+
+	if lines[0] == "STOPPING=1" {
+		return state.MySQLDown
+	}
+
+	if strings.HasPrefix(lines[0], "STATUS=") {
+		status := strings.TrimPrefix(lines[0], "STATUS=")
+		switch status {
+		case "Server is operational":
+			return state.MySQLReady
+		case "Server shutdown in progress":
+			return state.MySQLDown
+		case "Server startup in progress",
+			"Data Dictionary upgrade in progress",
+			"Data Dictionary upgrade complete",
+			"Server upgrade in progress",
+			"Server upgrade complete",
+			"Server downgrade in progress",
+			"Server downgrade complete",
+			"Data Dictionary upgrade from MySQL 5.7 in progress",
+			"Data Dictionary upgrade from MySQL 5.7 complete",
+			"Server shutdown complete": // we treat this as startup because during init, MySQL notifies this even if it's up
+			return state.MySQLStartup
+		}
+	}
+
+	return state.MySQLUnknown
+}
+
+func main() {
+	log.Println("Starting mysql-state-monitor")
+
+	socketPath, ok := os.LookupEnv(naming.EnvMySQLNotifySocket)
+	if !ok {
+		log.Fatalln("NOTIFY_SOCKET env variable is required")
+	}
+
+	stateFilePath, ok := os.LookupEnv(naming.EnvMySQLStateFile)
+	if !ok {
+		log.Fatalln("MYSQL_STATE_FILE env variable is required")
+	}
+
+	if _, err := os.Stat(socketPath); err == nil {
+		if err := os.Remove(socketPath); err != nil {
+			log.Fatalf("Failed to remove %s: %s", socketPath, err)
+		}
+	}
+
+	addr, err := net.ResolveUnixAddr("unixgram", socketPath)
+	if err != nil {
+		log.Fatalf("Failed to resolve unix addr %s: %s", socketPath, err)
+	}
+
+	conn, err := net.ListenUnixgram("unixgram", addr)
+	if err != nil {
+		log.Fatalf("Failed to listen unixgram %s: %s", socketPath, err)
+	}
+	defer conn.Close()
+
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		sig := <-sigterm
+		log.Printf("Received signal %v. Exiting mysql-state-monitor", sig)
+		os.Exit(0)
```

Comment:
When os.Exit(0) is executed, defers are not respected. This means that the `defer conn.Close()` we have in L.78 is not going to be executed and the connections will not be cleaned up properly. 

A workaround is to introduce a shutdown channel and when the sig is observed, the routine will close it. The main routine will for-select on this channel and return naturally while the default case will do what we currently do inside the already implemented for-loop. This is allow us to have defer always working as expected with this function.

If we want a faster workaround, we can just close connection inside the routine when the sig signal is observed, however future new defers will not work as expected in all scenarios.

Reply:
if process exits, connection will be closed anyway, no?

Reply:
Yes, from the Go process side, everything will be cleaned up, meaning any in-memory resources etc when the process exits, but the OS does not immediately clean up sockets on process termination If I'm not mistaken and it may take some time... during which time these resources still exist without reason.

Reply:
https://github.com/percona/percona-server-mysql-operator/pull/848/commits/35b5a1be3131d3cd6f355e44451ccdf8001e2ce2